### PR TITLE
fix(helm): return dry-run errors before release cleanup

### PR DESCRIPTION
**Pull request by raelga**
## Summary

Fix `helmdeploy` panic when the preliminary `helm upgrade --dry-run` fails before returning a release object.

This is the failure mode from ARO-HCP Ev2 fleet deployment logs:

```text
Failed to dry-run the Helm release: another operation (install/upgrade/rollback) is in progress
panic: runtime error: invalid memory address or nil pointer dereference
... removeOldFieldManager(... release *Release)
```

The deployer should return the original Helm error instead of trying to remove managed fields from a nil release.

## Changes

- Return immediately when the pre-deploy dry-run fails, preserving the original Helm error.
- Return rollout errors before converting a nil release, while still running diagnostics for real rollout failures.
- Treat nil Helm releases as explicit conversion errors.
- Add a regression test for nil release conversion.

## Validation

- `go test ./tools/helm`

Jira: [AROSLSRE-974](https://redhat.atlassian.net/browse/AROSLSRE-974) (sub-task of [AROSLSRE-975](https://redhat.atlassian.net/browse/AROSLSRE-975))

### CODE REVIEW

File: tools/helm/options.go [hunk 1]
```diff
@@ -310,6 +311,13 @@ func (opts *Options) Deploy(ctx context.Context) error {
 	releaser, releaseErr := runHelmUpgrade(ctx, logger, opts)
 	if releaseErr != nil {
 		logger.Error(releaseErr, "Failed to roll out the Helm release.")
+		if !opts.DryRun {
+			logger.Info("Running inline diagnostics.")
+			if err := runDiagnostics(ctx, logger, opts, deploymentStart); err != nil {
+				logger.Error(err, "Failed to capture diagnostics for Helm release")
+			}
+		}
```
[the remark this answers is not in the document]
Comment (raelga) [reply tools/helm/options.go]: Mitigated by https://github.com/Azure/ARO-Tools/pull/242/commits/5a9827e868d0d363985d8d3e17ca93c4995ecd43.

`runDiagnostics` now guards `release.Info` before logging status/description, and the rollout-error path returns the Helm error directly instead of invoking diagnostics with no release.

No further action needed.

File: tools/helm/options.go [hunk 2]
```diff
@@ -637,7 +642,7 @@ func releaserToV1Release(rel helmrelease.Releaser) (*helmreleasev1.Release, erro
 	case *helmreleasev1.Release:
 		return r, nil
 	case nil:
-		return nil, nil
+		return nil, errors.New("release is nil")
```
[the remark this answers is not in the document]
Comment (raelga) [reply tools/helm/options.go]: Mitigated by https://github.com/Azure/ARO-Tools/pull/242/commits/5a9827e868d0d363985d8d3e17ca93c4995ecd43.

The nil conversion error is now `helm releaser is nil`, and the unit test asserts that wording.

No further action needed.

File: tools/helm/options.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (raelga) [reply tools/helm/options.go]: Addressed by https://github.com/Azure/ARO-Tools/pull/242/commits/5a9827e868d0d363985d8d3e17ca93c4995ecd43.

I removed the new diagnostics-on-rollout-error behavior, so there is no new injected diagnostics execution path to test. Rollout failures now return the Helm error directly.

No further action needed.

File: tools/helm/options.go [hunk 3]
```diff
@@ -452,12 +451,18 @@ func runDiagnostics(ctx context.Context, logger logr.Logger, opts *Options, depl
...
 	logger.Info(
 		"Determined release status.",
 		"release", release.Name,
 		"namespace", release.Namespace,
-		"status", release.Info.Status,
-		"description", release.Info.Description,
+		"status", status,
+		"description", description,
 		"values", release.Config,
 	)
```
[the remark this answers is not in the document]
Comment (raelga) [reply tools/helm/options.go]: Mitigated by https://github.com/Azure/ARO-Tools/pull/242/commits/d82543b2fd76f62849e5d73be40230ddd26d651f.

The diagnostic log now includes an explicit `hasInfo` field, so zero-value status/description are distinguishable from a real populated `release.Info`.

No further action needed.

File: tools/helm/options.go [hunk 4]
```diff
@@ -637,7 +642,7 @@ func releaserToV1Release(rel helmrelease.Releaser) (*helmreleasev1.Release, erro
 	case *helmreleasev1.Release:
 		return r, nil
 	case nil:
-		return nil, nil
+		return nil, errors.New("helm releaser is nil")
```
[the remark this answers is not in the document]
Comment (raelga) [reply tools/helm/options.go]: Mitigated by https://github.com/Azure/ARO-Tools/pull/242/commits/d82543b2fd76f62849e5d73be40230ddd26d651f.

`releaserToV1Release(nil)` now returns the package-level sentinel `ErrNilHelmReleaser`, and the regression test checks it with `errors.Is`.

No further action needed.

File: tools/helm/options.go [hunk 5]
```diff
@@ -42,6 +42,9 @@ import (
 	"github.com/Azure/ARO-Tools/tools/cmdutils"
 )
 
+// ErrNilHelmReleaser is returned when Helm reports success without a release object.
+var ErrNilHelmReleaser = errors.New("helm releaser is nil")
```
[the remark this answers is not in the document]
Comment (raelga) [reply tools/helm/options.go]: Mitigated by https://github.com/Azure/ARO-Tools/pull/242/commits/51bf9c2b80a4de55d0ab431efe82239e87e092e4.

Renamed the sentinel to `ErrNilHelmReleaseResult` and changed the error text to `helm release result is nil`, so it describes Helm returning success without a release object.

No further action needed.

File: tools/helm/options.go [hunk 6]
```diff
@@ -452,12 +454,20 @@ func runDiagnostics(ctx context.Context, logger logr.Logger, opts *Options, depl
...
 		"Determined release status.",
 		"release", release.Name,
 		"namespace", release.Namespace,
-		"status", release.Info.Status,
-		"description", release.Info.Description,
+		"hasInfo", hasInfo,
+		"status", status,
+		"description", description,
 		"values", release.Config,
 	)
```
[the remark this answers is not in the document]
Comment (raelga) [reply tools/helm/options.go:472]: Mitigated by https://github.com/Azure/ARO-Tools/pull/242/commits/51bf9c2b80a4de55d0ab431efe82239e87e092e4.

When `release.Info` is nil, diagnostics now log `hasInfo=false`, `status=<missing>`, and `description=<missing>` instead of zero values.

No further action needed.

**Review by geoberle — APPROVED** [review]

**Review by geoberle — APPROVED** [review]
